subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/SnapshotCollectingDiffListener.java

```diff
@@ -36,10 +36,11 @@ public SnapshotCollectingDiffListener(Predicate<String> watchFilter) {
         addedSnapshots = new ArrayList<>();
     }
 
-    public void publishSnapshotDiff(SnapshotHierarchy.SnapshotDiffListener snapshotDiffListener) {
+    public SnapshotHierarchy publishSnapshotDiff(SnapshotHierarchy.SnapshotDiffListener snapshotDiffListener, SnapshotHierarchy newRoot) {
         if (!removedSnapshots.isEmpty() || !addedSnapshots.isEmpty()) {
-            snapshotDiffListener.changed(removedSnapshots, addedSnapshots);
+            return snapshotDiffListener.changed(removedSnapshots, addedSnapshots, newRoot);
         }
+        return newRoot;
```

Comment:
Why are we returning the parameter here?

Reply:
This allows the listener to have an effect on the VFS as well, e.g. when it fails to update the watchers. I'll check if I can move this out of the listener to somewhere else.

Reply:
I moved it to a separate error handling strategy, so the listener is as before.